internal/controller/deployment/controller_finalize.go

```diff
@@ -87,3 +129,48 @@ func (r *Reconciler) finalize(ctx context.Context, old, deployment *choreov1.Dep
 	}
 	return ctrl.Result{}, nil
 }
+
+func (r *Reconciler) cleanupEndpoints(ctx context.Context, deployment *choreov1.Deployment) error {
+	logger := log.FromContext(ctx).WithValues("deployment", deployment.Name)
+	logger.Info("Cleaning up the endpoints associated with the deployment")
+
+	// List all the endpoints associated with the deployment
+	endpointList := &choreov1.EndpointList{}
+	listOpts := []client.ListOption{
+		client.InNamespace(deployment.Namespace),
+		client.MatchingLabels{
+			labels.LabelKeyDeploymentName: deployment.Name,
+		},
+	}
+
+	if err := r.List(ctx, endpointList, listOpts...); err != nil {
+		if k8sapierrors.IsNotFound(err) {
+			logger.Info("No endpoints associated with the environment")
+			return nil
+		}
+		return fmt.Errorf("error listing endpoints: %w", err)
+	}
+
+	if len(endpointList.Items) == 0 {
+		logger.Info("No endpoints associated with the deployment")
+		return nil
+	}
+
+	for _, endpoint := range endpointList.Items {
+		// Check if the endpoint is being already deleting
+		if endpoint.DeletionTimestamp != nil {
+			continue
+		}
+
+		if err := r.Delete(ctx, &endpoint); err != nil {
+			if k8sapierrors.IsNotFound(err) {
+				// The endpoint is already deleted, no need to retry
+				continue
+			}
+			return fmt.Errorf("error deleting endpoint %s: %w", endpoint.Name, err)
+		}
+	}
+
+	// Reaching to this point means the endpoint deletion is still in progress.
+	return ErrEndpointDeletionWait
```

Comment:
We can directly use `(ctrl.Result, error)` return to indicate the reconcile result rather than relying on an error.

Reply:
Here we are returning `ErrEndpointDeletionWait` and catching at the top level to indicate that the deletion is still pending so that at the top level it is returning `(ctrl.Result, nil)` to avoid the immediate requeue. I'm expecting to trigger the next cycle once the endpoint is actually deleted through watch events.

Reply:
We can do the same by returning the `(ctrl.Result, error)` type, right?

Reply:
As per the offline discussion, will be using a (boolean, error).

Reply:
fixed